File: backend/app/api/v1/agents.py
"""Agent run endpoints + agent_log read endpoint.

Each signal agent has its own POST /agents/{name}/run for ad-hoc use.
POST /agents/analyze runs the full LangGraph workflow (all 5 agents in
parallel + Recommendation Agent + reflection).

  POST /agents/news/run
  POST /agents/technical/run
  POST /agents/fundamentals/run
  POST /agents/sentiment/run
  POST /agents/risk/run
  POST /agents/analyze       <- LangGraph multi-agent workflow
  GET  /agents/logs
"""
from __future__ import annotations

import logging

from fastapi import APIRouter, HTTPException, Query, status

from app.ai.agents.fundamentals_agent import (
    FundamentalsAgent,
    FundamentalsAgentInput,
)
from app.ai.agents.news_agent import NewsAgent, NewsAgentInput
from app.ai.agents.risk_agent import RiskAgent, RiskAgentInput
from app.ai.agents.sentiment_agent import SentimentAgent, SentimentAgentInput
from app.ai.agents.technical_agent import (
    TechnicalAgent,
    TechnicalAgentInput,
)
from app.api.deps import CurrentUser, DbDep
from app.schemas.agent import (
    AgentLogOut,
    FundamentalsAgentRunRequest,
    FundamentalsInsight,
    NewsAgentRunRequest,
    NewsInsight,
    RiskAgentRunRequest,
    RiskInsight,
    SentimentAgentRunRequest,
    SentimentInsight,
    TechnicalAgentRunRequest,
    TechnicalInsight,
)
from app.schemas.recommendation import AnalysisResult, AnalyzeRequest
from app.services import agent_log_service
from app.services.analysis_service import run_and_persist_analysis

log = logging.getLogger(__name__)
router = APIRouter()


# ---------- Single-agent runs ----------


@router.post(
    "/news/run",
    response_model=NewsInsight,
    summary="News Agent — fetches headlines, scores them",
)
async def run_news_agent(
    req: NewsAgentRunRequest, current_user: CurrentUser, db: DbDep
) -> NewsInsight:
    agent = NewsAgent()
    return await agent.run(
        NewsAgentInput(symbol=req.symbol, limit=req.limit),
        db=db,
        user_id=current_user.id,
    )


@router.post(
    "/technical/run",
    response_model=TechnicalInsight,
    summary="Technical Agent — pandas computes indicators, LLM interprets",
)
async def run_technical_agent(
    req: TechnicalAgentRunRequest, current_user: CurrentUser, db: DbDep
) -> TechnicalInsight:
    agent = TechnicalAgent()
    return await agent.run(
        TechnicalAgentInput(
            symbol=req.symbol,
            range_=req.range_,
            interval=req.interval,
        ),
        db=db,
        user_id=current_user.id,
    )


@router.post(
    "/fundamentals/run",
    response_model=FundamentalsInsight,
    summary="Fundamentals Agent — ratios + RAG over filings",
)
async def run_fundamentals_agent(
    req: FundamentalsAgentRunRequest, current_user: CurrentUser, db: DbDep
) -> FundamentalsInsight:
    agent = FundamentalsAgent()
    return await agent.run(
        FundamentalsAgentInput(symbol=req.symbol, use_rag=req.use_rag),
        db=db,
        user_id=current_user.id,
    )


@router.post(
    "/sentiment/run",
    response_model=SentimentInsight,
    summary="Sentiment Agent — Reddit crowd-mood",
)
async def run_sentiment_agent(
    req: SentimentAgentRunRequest, current_user: CurrentUser, db: DbDep
) -> SentimentInsight:
    agent = SentimentAgent()
    return await agent.run(
        SentimentAgentInput(symbol=req.symbol, limit=req.limit),
        db=db,
        user_id=current_user.id,
    )


@router.post(
    "/risk/run",
    response_model=RiskInsight,
    summary="Risk Agent — volatility, drawdown, leverage",
)
async def run_risk_agent(
    req: RiskAgentRunRequest, current_user: CurrentUser, db: DbDep
) -> RiskInsight:
    agent = RiskAgent()
    return await agent.run(
        RiskAgentInput(symbol=req.symbol, range_=req.range_),
        db=db,
        user_id=current_user.id,
    )


# ---------- Multi-agent orchestration ----------


@router.post(
    "/analyze",
    response_model=AnalysisResult,
    summary=(
        "Run the full LangGraph workflow: 5 signal agents in parallel + "
        "Recommendation synthesis + reflection loop."
    ),
)
async def analyze_symbol(
    req: AnalyzeRequest, current_user: CurrentUser, db: DbDep
) -> AnalysisResult:
    try:
        rec = await run_and_persist_analysis(
            user_id=current_user.id, symbol=req.symbol, db=db
        )
    except RuntimeError as e:
        # All agents failed or Recommendation node failed.
        raise HTTPException(
            status_code=status.HTTP_502_BAD_GATEWAY,
            detail={"message": "Workflow produced no recommendation", "error": str(e)},
        )
    except Exception as e:
        log.exception("Workflow failed")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Workflow failed: {e}",
        )
    return AnalysisResult.from_recommendation(rec)


# ---------- Logs ----------


@router.get(
    "/logs",
    response_model=list[AgentLogOut],
    summary="Recent agent runs (success + failure) for the current user",
)
async def list_logs(
    current_user: CurrentUser,
    db: DbDep,
    agent: str | None = Query(default=None, description="Filter by agent name"),
    symbol: str | None = Query(default=None, description="Filter by symbol"),
    limit: int = Query(default=50, ge=1, le=200),
) -> list[AgentLogOut]:
    rows = await agent_log_service.list_for_user(
        db, current_user, agent=agent, symbol=symbol, limit=limit
    )
    return [AgentLogOut.model_validate(r) for r in rows]

File: backend-ts/src/routes/agents.ts
/**
 * Agent endpoints.
 *
 *   POST /agents/:name/run     run one agent (news|technical|fundamentals|sentiment|risk)
 *   POST /agents/analyze       run the full multi-agent workflow (Phase 7)
 *   GET  /agents/logs          recent runs across all agents
 */
import type { FastifyInstance } from "fastify";
import { z } from "zod";
import { prisma } from "../db.js";
import { requireAuth } from "../middleware/auth.js";
import { agents, type AgentName } from "../ai/agents/registry.js";
import { runWorkflow } from "../ai/workflow/graph.js";

const runSchema = z.object({
  symbol: z.string().min(1).max(20),
  limit: z.number().int().positive().optional(),
  range: z.string().optional(),
  interval: z.string().optional(),
  use_rag: z.boolean().optional(),
});

const analyzeSchema = z.object({
  symbol: z.string().min(1).max(20),
});

export default async function agentRoutes(app: FastifyInstance) {
  app.addHook("preHandler", requireAuth);

  app.post<{ Params: { name: string } }>("/:name/run", async (req, reply) => {
    const parsed = runSchema.safeParse(req.body);
    if (!parsed.success) return reply.code(422).send({ detail: parsed.error.flatten() });
    const name = req.params.name as AgentName;
    const agent = agents[name];
    if (!agent) return reply.code(404).send({ detail: `Unknown agent: ${name}` });
    try {
      const result = await agent.run({
        ...parsed.data,
        symbol: parsed.data.symbol.toUpperCase(),
        userId: req.currentUser!.id,
      });
      return { symbol: parsed.data.symbol.toUpperCase(), ...result.output };
    } catch (e) {
      app.log.error(e);
      return reply.code(502).send({ detail: (e as Error).message });
    }
  });

  app.post("/analyze", async (req, reply) => {
    const parsed = analyzeSchema.safeParse(req.body);
    if (!parsed.success) return reply.code(422).send({ detail: parsed.error.flatten() });
    try {
      const result = await runWorkflow(parsed.data.symbol.toUpperCase(), req.currentUser!.id);
      return result;
    } catch (e) {
      app.log.error(e);
      return reply.code(502).send({ detail: (e as Error).message });
    }
  });

  app.get<{
    Querystring: { agent?: string; symbol?: string; limit?: string };
  }>("/logs", async (req) => {
    const where: Record<string, unknown> = { userId: req.currentUser!.id };
    if (req.query.agent) where.agentName = req.query.agent;
    if (req.query.symbol) where.symbol = req.query.symbol.toUpperCase();
    const limit = Math.min(Number(req.query.limit) || 50, 200);
    const rows = await prisma.agentLog.findMany({
      where: where as never,
      orderBy: { createdAt: "desc" },
      take: limit,
    });
    return rows.map((r) => ({
      id: r.id,
      agent_name: r.agentName,
      symbol: r.symbol,
      status: r.status,
      duration_ms: r.durationMs,
      error: r.error,
      provider: r.provider,
      model: r.model,
      input_data: r.inputData,
      output_data: r.outputData,
      created_at: r.createdAt.toISOString(),
    }));
  });
}
